### PR TITLE
P2P clarifications when introducing `engine_getBlobsV1`

### PR DESCRIPTION
**Pull request by tbenr**
This PR adds some clarifications in the P2P spec if [engine_getBlobsV1](https://github.com/ethereum/execution-apis/pull/559) become a viable method to retrieve blobs.

Dictates that an honest validator, when successfully recover blobs for the current slot using local EL MUST:
- republish the `blob_sidecar` over the p2p layer
- make sure it updates all the data structures as if client had received the `blob_sidecar` over the gossip (ie update equivocation cache)

The basic idea is that, if `engine_getBlobsV1` is used to recover missing blobs while following the head, it should behave like it has been received via gossip.
If it is used to retrieve recent past blobs the node could have missed, it should act as a byRoot RPC request (no republish)

### CODE REVIEW

File: specs/deneb/p2p-interface.md [hunk 1]
```diff
@@ -295,7 +296,7 @@ Before consuming the next response chunk, the response reader SHOULD verify the
 
 No more than `MAX_REQUEST_BLOB_SIDECARS` may be requested at a time.
 
-`BlobSidecarsByRoot` is primarily used to recover recent blobs (e.g. when receiving a block with a transaction whose corresponding blob is missing).
+`BlobSidecarsByRoot` is primarily used to recover recent blobs (e.g. when receiving a block with a transaction whose corresponding blob is missing). In addition, it is possible to retrieve recent blobs by querying the Execution Layer via `engine_getBlobsV1` engine API method.
```
Comment (dapplion) [inline specs/deneb/p2p-interface.md:299]: Does this sentence mean that an external peer can trigger the node into doing an EL query?

**Comment by mkalinin** [comment]
> So the question is: how bad would be to mostly rely on EL mempool for data availability? what are the risks?

The risk is that nodes that aren’t connected to EL mempool or those that didn’t receive expected blobs on EL for any reason won’t be able to conclude DA. Probably, they will send `ByRoot` request and a remote peer will send those blobs retrieved via EL, but if neither remote peer has those blobs in the EL mempool then it will delay DA consideration. And we will likely see a split in attestations. I am not sure if any practical attack becomes possible because of that, but looks like heavily relying on EL mempool can increase attacking surface.

File: specs/deneb/p2p-interface.md (hunk 1 above)
Comment (tbenr) [reply specs/deneb/p2p-interface.md:299]: I see what you mean. Probably worth better clarifying, but the context here is on the "consumer" of the byRoot, so my intent was to mention this new method that can be used  as alternative (or in combination) to p2p RPC for recovering recent blobs.
EL query triggered by RCP seems to me an edge case, theoretically possible, I'd leave it as an implementation detail.

**Comment by jimmygchen** [comment]
> So the question is: how bad would be to mostly rely on EL mempool for data availability? what are the risks?

I think we can't just rely on the EL mempool for data availability. We would still require them to be broadcasted if the blob txs come from private mempools. Although viewing it as a complimentary thing is super useful:
- **Reduces block import latency**: nodes can retrieve blobs from EL without waiting for them from gossip, hence making _some_ blocks attestable earlier.
- **Improves network resiliency**: being able to fetch blobs from EL reduces the likelihood of missed block due to delays in gossip propagation.
- **Allows further scaling without sacrificing decentralization**: nodes with more resources _could_ participate in sidecar computation and propagation, allowing nodes with limited bandwidth to continue to produce block post-PeerDAS.

We've implemented an [experimental integration](https://github.com/sigp/lighthouse/pull/6268) with this endpoint in Lighthouse, and it's proven to be quite reliable on testnets, and is probably the simplest and immediate solution to proposer bandiwdth problem after Pectra / PeerDAS.

**Comment by mkalinin** [comment]
@tbenr what do you think about using “request from local execution layer client” in lieu of a specific Engine API method? The spec can mention a specific method as an example of a way of direct EL querying. In general, the spec is abstracted away from Engine API and its spec

**Comment by tbenr** [comment]
> you think about using “request from local execution layer client” in lieu 

that's make sense!

File: specs/deneb/p2p-interface.md [hunk 2]
```diff
@@ -184,6 +185,7 @@ The following validations MUST pass before forwarding the `blob_sidecar` on the
 - _[REJECT]_ The sidecar's inclusion proof is valid as verified by `verify_blob_sidecar_inclusion_proof(blob_sidecar)`.
 - _[REJECT]_ The sidecar's blob is valid as verified by `verify_blob_kzg_proof(blob_sidecar.blob, blob_sidecar.kzg_commitment, blob_sidecar.kzg_proof)`.
 - _[IGNORE]_ The sidecar is the first sidecar for the tuple `(block_header.slot, block_header.proposer_index, blob_sidecar.index)` with valid header signature, sidecar inclusion proof, and kzg proof.
+*Note:* If client obtains the blob via local execution layer client or any other method before receiving the corresponding sidecar via this topic, it MUST NOT consider the tuple as "seen" (i.e. it MUST participate in the dissemination of the blob_sidecar even if it already obtained the blob)
```
Comment (tbenr) [inline specs/deneb/p2p-interface.md:188]: @mkalinin not sure about this specification now that there is a requirement to publish the `blob_sidecar` whenever is recovered.
WDYT?

Comment (mkalinin) [reply specs/deneb/p2p-interface.md:188]: I think it makes sense to mention how to deal with the sidecar gossip in the EL blob recovery section. i.e. mention that if the sidecar has been published after EL blob recovery then it must also be considered as seen.

Comment (tbenr) [reply specs/deneb/p2p-interface.md]: I updated it

File: specs/deneb/p2p-interface.md [hunk 3]
```diff
@@ -309,6 +310,12 @@ Clients SHOULD include a sidecar in the response as soon as it passes the gossip
 Clients SHOULD NOT respond with sidecars related to blocks that fail gossip validation rules.
 Clients SHOULD NOT respond with sidecars related to blocks that fail the beacon chain state transition
 
+###### Blobs recovery via local execution layer client
+In addition to `BlobSidecarsByRoot` requests, recent blobs recovery MAY also be done by querying the Execution Layer (i.e. via `engine_getBlobsV1`)
+Implementers are encouraged to leverage this method to increase the likelihood of incorporating and attesting to the last block when its proposer is not able to publish blobs on time.
+
+Clients MUST publish the corresponding `blob_sidecar` on the `blob_sidecar_{subnet_id}` subnet whenever it successfully recovers blobs via local execution layer client. In such case, clients MUST also consider the `blob_sidecar` as "seen" with regards to blob subnet gossip rules.
```
Comment (dapplion) [inline specs/deneb/p2p-interface.md:317]: When you publishing a message to gossip it gets included in the seen cache right? Why is it necessary to add this sentence here?

> clients MUST also consider the `blob_sidecar` as "seen" with regards to blob subnet gossip rules

Comment (tbenr) [reply specs/deneb/p2p-interface.md:317]: I'm not referring to libp2p "seen" cache here but rather the equivocation tracking tuple of the gossip rule. I'll revisit it

Comment (tbenr) [reply specs/deneb/p2p-interface.md:317]: Thinking about adding `(i.e "the sidecar is the first sidecar for the tuple ..." rule)` but doesn't look good to me. Any suggestion on how to improve? Assuming you are ok with the semantics.

Comment (tbenr) [reply specs/deneb/p2p-interface.md:317]: maybe something like:

```
Clients MUST also update their data structures as if the `blob_sidcar` had been received via gossip (ie updating the anti-equivocation caches)
```

Comment (dapplion) [reply specs/deneb/p2p-interface.md]: Thanks! The clarification sounds better to me

**Review by dapplion — APPROVED** [review]

**Review by mkalinin — APPROVED** [review]

**Review by jimmygchen — APPROVED** [review]
Nice, looks great!

File: specs/deneb/p2p-interface.md [hunk 4]
```diff
@@ -309,6 +310,14 @@ Clients SHOULD include a sidecar in the response as soon as it passes the gossip
 Clients SHOULD NOT respond with sidecars related to blocks that fail gossip validation rules.
 Clients SHOULD NOT respond with sidecars related to blocks that fail the beacon chain state transition
 
+###### Blobs recovery via local execution layer client
+In addition to `BlobSidecarsByRoot` requests, recent blobs recovery MAY also be done by querying the Execution Layer (i.e. via `engine_getBlobsV1`)
+Implementers are encouraged to leverage this method to increase the likelihood of incorporating and attesting to the last block when its proposer is not able to publish blobs on time.
+
+When clients uses local execution layer to recover missing blobs relative to the current slot, they MUST behave as if the corresponding `blob_sidecar` had been received via gossip. In particular they MUST:
```
Comment (mehdi-aouadi) [inline specs/deneb/p2p-interface.md:317]:
```suggestion
When clients use local execution layer to recover missing blobs
```

Comment (mehdi-aouadi) [inline specs/deneb/p2p-interface.md:317]:
```suggestion
When clients use local execution layer to recover missing blobs relative to the current slot, they MUST behave as if the corresponding `blob_sidecar` had been received via gossip. In particular they MUST:
```

**Review by mehdi-aouadi — APPROVED** [review]

File: specs/deneb/p2p-interface.md [hunk 5]
```diff
@@ -34,6 +34,7 @@ The specification of these changes continues in the same format as the network s
       - [BeaconBlocksByRange v2](#beaconblocksbyrange-v2)
       - [BeaconBlocksByRoot v2](#beaconblocksbyroot-v2)
       - [BlobSidecarsByRoot v1](#blobsidecarsbyroot-v1)
+        - [Blobs recovery via local execution layer client](#blobs-recovery-via-local-execution-layer-client)
```
Comment (jtraglia) [inline specs/deneb/p2p-interface.md:37]:
```suggestion
        - [Blob retrieval via local execution layer client](#blob-retrieval-via-local-execution-layer-client)
```

**Review by jtraglia — APPROVED** [review]
Looks great. Thanks for making those changes 🙏

**Comment by Scutua** [comment]
Everything ok..? I’m available around this time everyday. Have to sorry for late response.

**Pull request merged**